libraries/botbuilder-core/src/adapterExtensions.ts

```diff
@@ -0,0 +1,28 @@
+/**
+ * @module botbuilder
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { BotAdapter } from './botAdapter';
+import { BotState } from './botState';
+import { RegisterClassMiddleware } from './registerClassMiddleware';
+
+export class AdapterExtensions {
+    /**
+     * Adds middleware to the adapter to register one or more BotState objects on the turn context.
+     * The middleware registers the state objects on the turn context at the start of each turn.
+     * @param botAdapter The adapter on which to register the state objects.
+     * @param botStates The state objects to register.
+     */
+    public static useBotState(botAdapter: BotAdapter, ...botStates: BotState[]): BotAdapter {
+        for (const botState of botStates) {
+            const key = botState.constructor.name;
```

Comment:
What about name collisions? Any concern there? Or will users never try to register multiple states derived from the same class?

Reply:
We may have multiple `State` instances extending `BotState` like `UserState` and `ConversationState`, but there is only one instance for each kind of state in a dialog manager, which will be consumed at the `onTurn()` method of `DialogManager` or somewhere else.